lib/helpers/cursor/eachAsync.js

```diff
@@ -66,8 +70,25 @@ module.exports = function eachAsync(next, fn, options, callback) {
         // make sure we know that we still have a result to handle re: #8422
         process.nextTick(() => done());
 
-        handleNextResult(doc, currentDocumentIndex++, function(err) {
-          --handleResultsInProgress;
+        if (batchSize) {
+          documentsBatch.push(doc);
+        }
+
+        // If the current documents size is less than the provided patch size don't process the documents yet
+        if (batchSize && documentsBatch.length !== batchSize) {
```

Comment:
I'd do `documentsBatch.length < batchSize` just in case, but I don't see that being a blocker here.